### PR TITLE
[Feature](partial update) Support flexible partial update in stream load with json files

### PR DESCRIPTION
**Pull request by bobhan1**
This PR add the ability to update different columns for each row in one stream load
Doc: https://github.com/apache/doris-website/pull/1140
## Example
```sql
MySQL root@127.1:d1> CREATE TABLE t1 (
                  -> `k` int(11) NULL, 
                  -> `v1` BIGINT NULL,
                  -> `v2` BIGINT NULL DEFAULT "9876",
                  -> `v3` BIGINT NOT NULL,
                  -> `v4` BIGINT NOT NULL DEFAULT "1234",
                  -> `v5` BIGINT NULL
                  -> ) UNIQUE KEY(`k`) DISTRIBUTED BY HASH(`k`) BUCKETS 1
                  -> PROPERTIES(
                  -> "replication_num" = "1",
                  -> "enable_unique_key_merge_on_write" = "true"); 
Query OK, 0 rows affected
Time: 0.013s
MySQL root@127.1:d1> insert into t1 select number, number, number, number, number, number from numbers("number" = "6");
Query OK, 6 rows affected
Time: 0.107s
MySQL root@127.1:d1> select * from t1;
+---+----+----+----+----+----+
| k | v1 | v2 | v3 | v4 | v5 |
+---+----+----+----+----+----+
| 0 | 0  | 0  | 0  | 0  | 0  |
| 1 | 1  | 1  | 1  | 1  | 1  |
| 2 | 2  | 2  | 2  | 2  | 2  |
| 3 | 3  | 3  | 3  | 3  | 3  |
| 4 | 4  | 4  | 4  | 4  | 4  |
| 5 | 5  | 5  | 5  | 5  | 5  |
+---+----+----+----+----+----+
```
test1.json:
```json
{"k": 1, "v1": 10}
{"k": 2, "v2": 20, "v5": 25}
{"k": 3, "v3": 30}
{"k": 4, "v4": 20, "v1": 43, "v3": 99}
{"k": 5, "v5": null}
{"k": 6, "v1": 999, "v3": 777}
{"k": 2, "v4": 222}
{"k": 1, "v2": 111, "v3": 111}
```

```bash
curl --location-trusted -u root: \
-H "strict_mode:false" \
-H "format:json" \
-H "read_json_by_line:true" \
-H "unique_key_update_mode:UPDATE_FLEXIBLE_COLUMNS" \
-T test1.json \
-XPUT http://<host>:<http_port>/api/d1/t1/_stream_load
```

```sql
MySQL root@127.1:d1> select * from t1;
+---+-----+------+-----+------+--------+
| k | v1  | v2   | v3  | v4   | v5     |
+---+-----+------+-----+------+--------+
| 0 | 0   | 0    | 0   | 0    | 0      |
| 1 | 10  | 111  | 111 | 1    | 1      |
| 2 | 2   | 20   | 2   | 222  | 25     |
| 3 | 3   | 3    | 30  | 3    | 3      |
| 4 | 43  | 4    | 99  | 20   | 4      |
| 5 | 5   | 5    | 5   | 5    | <null> |
| 6 | 999 | 9876 | 777 | 1234 | <null> |
+---+-----+------+-----+------+--------+
```

### CODE REVIEW

File: gensrc/thrift/Descriptors.thrift [hunk 1]
```diff
@@ -248,12 +248,14 @@ struct TOlapTableSchemaParam {
...
     7: optional bool is_dynamic_schema // deprecated
-    8: optional bool is_partial_update
+    8: optional bool is_partial_update // deprecated, use unique_key_update_mode
     9: optional list<string> partial_update_input_columns
     10: optional bool is_strict_mode = false
     11: optional string auto_increment_column
     12: optional i32 auto_increment_column_unique_id = -1
     13: optional Types.TInvertedIndexFileStorageFormat inverted_index_file_storage_format = Types.TInvertedIndexFileStorageFormat.V1
+    14: optional Types.TUniqueKeyUpdateMode unique_key_update_mode = Types.TUniqueKeyUpdateMode.UPSERT
+    15: optional i32 sequence_map_col_unique_id = -1
```
Comment (dataroaring) [inline gensrc/thrift/Descriptors.thrift:258]: Maybe merge_bitmap is a better name.

File: gensrc/proto/olap_file.proto [hunk 2]
```diff
@@ -390,6 +396,7 @@ message TabletSchemaPB {
     // column unique ids for row store columns
     repeated int32 row_store_column_unique_ids = 26;
     optional int64 row_store_page_size = 27 [default=16384];
+    optional int32 skip_bitmap_col_idx = 29 [default = -1];
```
[the remark this answers is not in the document]
Comment (dataroaring) [reply gensrc/proto/olap_file.proto:399]: CloudTabletSchemaPB should be handled.

[the remark this answers is not in the document]
Comment (bobhan1) [reply gensrc/proto/olap_file.proto:399]: > what about 28?

@dataroaring 28 is used in `TabletSchemaCloudPB`

File: be/src/exec/tablet_info.cpp [hunk 3]
```diff
@@ -186,11 +199,42 @@ Status OlapTableSchemaParam::init(const TOlapTableSchemaParam& tschema) {
...
+            _sequence_map_col_uid = tschema.sequence_map_col_unique_id;
+        }
+    } else {
+        // for backward compatibility
+        if (tschema.__isset.is_partial_update && tschema.is_partial_update) {
+            _unique_key_update_mode = UniqueKeyUpdateModePB::UPDATE_FIXED_COLUMNS;
+        } else {
+            _unique_key_update_mode = UniqueKeyUpdateModePB::UPSERT;
+        }
+    }
```
Comment (dataroaring) [inline be/src/exec/tablet_info.cpp:233]: put into a new function.

File: be/src/olap/base_tablet.cpp [hunk 4]
```diff
@@ -983,12 +994,111 @@ Status BaseTablet::generate_new_block_for_partial_update(
...
+                if (skip_bitmaps->at(idx).contains(col_uid)) {
+                    if (old_block_delete_signs != nullptr &&
+                        old_block_delete_signs[read_index_old[idx]] != 0) {
+                        if (rs_column.has_default_value()) {
+                            const auto& src_column =
+                                    *default_value_block
+                                             .get_by_position(cid -
+                                                              rowset_schema->num_key_columns())
+                                             .column;
+                            cur_col->insert_from(src_column, 0);
```
Comment (dataroaring) [inline be/src/olap/base_tablet.cpp:1080]: too many idents. Please consider put some code in a function.

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run cloud_p0

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Review by zhannngchen — APPROVED** [review]
LGTM

**Comment by Carl-Zhou-CN** [comment]
@dataroaring @zhannngchen hi, is there any plan to release this PR? What risks are involved

**Comment by bobhan1** [comment]
@Carl-Zhou-CN the feature will probably be in doris 4.0

**Comment by Carl-Zhou-CN** [comment]
> @Carl-Zhou-CN the feature will probably be in doris 4.0

Thank you very much for your response. I currently need this functionality urgently. If I merge and use it myself, what issues should I be aware of?

**Comment by bobhan1** [comment]
> > @Carl-Zhou-CN the feature will probably be in doris 4.0
> 
> Thank you very much for your response. I currently need this functionality urgently. If I merge and use it myself, what issues should I be aware of?

This PR has problems and some them are fixed in https://github.com/apache/doris/pull/41701.
This feature may require high iops and io throughput.

**Comment by Carl-Zhou-CN** [comment]
> > > @Carl-Zhou-CN the feature will probably be in doris 4.0
> > 
> > 
> > Thank you very much for your response. I currently need this functionality urgently. If I merge and use it myself, what issues should I be aware of?
> 
> This PR has problems and some them are fixed in #41701. This feature may require high iops and io throughput.

Ok. Thank you.

**Pull request merged**